hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/producer/kafka/KafkaMessagesAsyncProducer.java

```diff
@@ -0,0 +1,86 @@
+package pl.allegro.tech.hermes.frontend.producer.kafka;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import pl.allegro.tech.hermes.api.Topic;
+import pl.allegro.tech.hermes.frontend.metric.CachedTopic;
+import pl.allegro.tech.hermes.frontend.producer.BrokerMessageProducer;
+import pl.allegro.tech.hermes.frontend.producer.BrokerMessageProducingResult;
+import pl.allegro.tech.hermes.frontend.producer.BrokerMessagesAsyncProducer;
+import pl.allegro.tech.hermes.frontend.producer.BrokerMessagesAsyncProducingObserver;
+import pl.allegro.tech.hermes.frontend.producer.BrokerMessagesProducingResults;
+import pl.allegro.tech.hermes.frontend.publishing.PublishingCallback;
+import pl.allegro.tech.hermes.frontend.publishing.message.Message;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.validation.constraints.NotNull;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static java.util.concurrent.CompletableFuture.runAsync;
+
+@Singleton
+public class KafkaMessagesAsyncProducer implements BrokerMessagesAsyncProducer {
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaMessagesAsyncProducer.class);
+
+    private final BrokerMessageProducer brokerMessageProducer;
+    private final ExecutorService executorService;
+
+    @Inject
+    public KafkaMessagesAsyncProducer(BrokerMessageProducer brokerMessageProducer, int threadPoolSize) {
+        this.brokerMessageProducer = brokerMessageProducer;
+        this.executorService = Executors.newFixedThreadPool(
+                threadPoolSize,
+                new ThreadFactoryBuilder().setNameFormat("kafka-messages-async-publishing-%d").build()
+        );
+    }
+
+    @Override
+    public BrokerMessagesProducingResults publishMessagesAsync(@NotNull CachedTopic topic, List<Message> messages, long timeoutMs) {
+        if (messages.size() > 0) {
+            final BrokerMessagesAsyncProducingObserver observer = new KafkaMessagesAsyncProducingObserver(messages.size(), timeoutMs);
+            messages.forEach(it -> unwrapAsync(() -> publishMessage(topic, it, observer)));
```

Comment:
If you use single threaded executor you don't need any async-like code, messages will be sent sequentially and it will simplifify this code :)

Reply:
now sending is sequential